Fix race condition for wasm referenced functions

hyjorc1 opened the pull request:
#### 527e2a01d73c94cc018b1fa060fdffc15bab69d6
<pre>
Fix race condition for wasm referenced functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=250847">https://bugs.webkit.org/show_bug.cgi?id=250847</a>
rdar://104332636

Reviewed by Mark Lam.

Fix race condition for updating referenced functions while compiling
wasm functions concurrently.

* JSTests/wasm/stress/referenced-function.js: Added.
(async let):
(j.catch):
* JSTests/wasm/stress/resources/funcref-race.wasm: Added.
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
(JSC::Wasm::ModuleInformation::referencedFunctions const):
(JSC::Wasm::ModuleInformation::hasReferencedFunction const):
(JSC::Wasm::ModuleInformation::addReferencedFunction const):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FixedBitVector.h: Added.
(WTF::FixedBitVector::concurrentAllocateOnce):
(WTF::FixedBitVector::concurrentTest):
(WTF::FixedBitVector::concurrentTestAndSet):
(WTF::FixedBitVector::testAndSet):
(WTF::FixedBitVector::test):
(WTF::FixedBitVector::findBit const):
(WTF::FixedBitVector::operator== const):
(WTF::FixedBitVector::hash const):
(WTF::FixedBitVector::get const):
(WTF::FixedBitVector::dump const):
(WTF::FixedBitVectorHash::hash):
(WTF::FixedBitVectorHash::equal):
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/FixedBitVector.cpp: Added.
(TestWebKitAPI::testFixedBitVectorSize):
(TestWebKitAPI::testFixedBitVectorTest):
(TestWebKitAPI::testFixedBitVectorTestAndSet):
(TestWebKitAPI::testFixedBitVectorConcurrentTest):
(TestWebKitAPI::testFixedBitVectorConcurrentTestAndSet):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/259323@main">https://commits.webkit.org/259323@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7ee36e81ea975b639384d6176d965cefeb02755

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104642 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13716 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/37543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113917 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108559 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14841 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4647 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/96972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/112857 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110405 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/37543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/96972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/5/builds/108114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/37543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/96972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/94600 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7074 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/37543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92534 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/83/builds/4842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/7181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/30104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/20/builds/103471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/13228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/37543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101220 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6429 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/8970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/25173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (changes requested):
We discussed internally, and we should make this BitVector updated lock-free.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmModuleInformation.h:
```diff
@@ -91,8 +91,10 @@ struct ModuleInformation : public ThreadSafeRefCounted<ModuleInformation> {
-    const BitVector& referencedFunctions() const { return m_referencedFunctions; }
-    void addReferencedFunction(unsigned index) const { m_referencedFunctions.set(index); }
+    void allocateReferencedFunctions() const { m_referencedFunctions.allocateOnce(functionIndexSpaceSize()); }
+    const FixedBitVector& referencedFunctions() const { return m_referencedFunctions; }
```
Is this still used?  If not, let's remove it.

hyjorc1 replied:
Yes, it's used in `void WebAssemblyModuleRecord::initializeExports(JSGlobalObject* globalObject)` of `WebAssemblyModuleRecord.cpp`.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmModuleInformation.h:
```diff
@@ -91,8 +91,10 @@ struct ModuleInformation : public ThreadSafeRefCounted<ModuleInformation> {
 
-    const BitVector& referencedFunctions() const { return m_referencedFunctions; }
-    void addReferencedFunction(unsigned index) const { m_referencedFunctions.set(index); }
+    void allocateReferencedFunctions() const { m_referencedFunctions.allocateOnce(functionIndexSpaceSize()); }
```
I suggest naming this `initializeReferencedFunctionsTracker` (or something like that) instead.  `allocateReferencedFunctions` suggests that we're allocating memory for the functions themselves somehow, but this isn't doing anything of the sort.  What it does is initialize the `m_referencedFunctions` vector, which serves as a tracker for whether a function is referenced or not ... hence the name I suggested.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
```diff
@@ -232,6 +232,9 @@ auto SectionParser::parseFunction() -> PartialResult
         m_info->functions.uncheckedAppend({ start, end, false, false, Vector<uint8_t>() });
     }
 
+    // Note that `allocateReferencedFunctions` can be used after both parseImport and parseFunction
```
/can be used/should only be called/

MenloDorian commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,300 @@
+namespace WTF {
+
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(FixedBitVector);
+DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(BitVector);
```
This is not needed, right?  Please remove.

Constellation commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,300 @@
+    bool concurrentTest(size_t bitIndex, Dependency = Dependency());
+    bool concurrentTestAndSet(size_t bitIndex, Dependency = Dependency());
+
+    void allocateOnce(size_t);
```
Let's not modify the size after it is constructed (so let's not add allocate method)
We should take size in the constructor, and we should use it like "vector = FixedBitVector(42)"

MenloDorian commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,300 @@
+    m_size = size;
+    m_words = (size + wordSize - 1) / wordSize;
+    m_bits = (WordType*)FixedBitVectorMalloc::tryZeroedMalloc(m_words * sizeof(WordType));
+}
```
These can be removed based on Yusuke's feedback.

Constellation commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,300 @@
+
+    size_t m_size { 0 };
+    unsigned m_words { 0 };
+    WordType* m_bits { nullptr };
```
I think we should use FixedVector for this storage. Then we can use m_bits.size() instead of m_words here. And it reduces the size of this class. Plus, it simplifies the implementation a lot. We do not need custom destructor (since FixedVector destructor destroys it). We do not need to have copy constructors since FixedVector will be copied. Probably we can make operator== simpler too.

MenloDorian commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,300 @@
+        [&](WordType& value) -> bool {
+            return value & mask;
+        });
+}
```
It doesn't make sense to have a `concurrentTest`.  This does not fix and race in the client.  An we should not have any race internally because we should never be reallocating the backing store.  Remove this.

MenloDorian commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,300 @@
+    }
+
+    bool concurrentAllocateOnce(size_t);
+    bool concurrentTest(size_t bitIndex, Dependency = Dependency());
```
Remove these.

MenloDorian commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,300 @@
+
+    bool concurrentAllocateOnce(size_t);
+    bool concurrentTest(size_t bitIndex, Dependency = Dependency());
+    bool concurrentTestAndSet(size_t bitIndex, Dependency = Dependency());
```
I think it's worth implementing a `concurrentTestAndClear` even though we don't use it yet.

MenloDorian commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,300 @@
+    bool concurrentTestAndSet(size_t bitIndex, Dependency = Dependency());
+
+    void allocateOnce(size_t);
+    bool testAndSet(size_t bitIndex);
```
Also add `testAndClear`.

MenloDorian commented on Tools/TestWebKitAPI/Tests/WTF/Bitmap.cpp:
```diff
@@ -271,13 +271,13 @@ void testBitmapConcurrentTestAndClear()
         ASSERT_EQ(bitmap1.get(i), expectedBits1[i]);
     for (size_t i = 0; i < size; ++i)
-        ASSERT_EQ(bitmap1.testAndClear(i), expectedBits1[i]);
+        ASSERT_EQ(bitmap1.concurrentTestAndClear(i), expectedBits1[i]);
```
The same bug also exists in `testBitmapConcurrentTestAndSet`.  Please fix that also.

MenloDorian commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,238 @@
+    static constexpr unsigned wordSize = sizeof(WordType) * 8;
+    static constexpr WordType one = 1;
+
+    size_t m_size { 0 };
```
`BitVector` has 1 neat trick: it encodes all the bits into a `uintptr_t` if the requested number of bits (minus 1) will in the `uintptr_t` itself.  This creates a quirk: if the client requests less bits than that, the `BitVector` still gives it the max inlined bits for the size.  So, one advantage of using a `BitVector` as the backing store is that it makes `FixedBitVector` have the same matching quirk.  This means `FixedBitVector` gets to be a drop in replacement for BitVector, and other than the non-resizable property, we can expect it to work the same way.

But in order to inherit that quirk, we'll have to use `BitVector`'s `size()` tracking instead of rolling our own with this `m_size` field.  Additionally, since `BitVector` already has a means of tracking `size`, it is inefficient to separately track it here.  So, let's remove this `m_size` field.  It is not needed.

MenloDorian commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,238 @@
+    FixedBitVector() = default;
+
+    FixedBitVector(size_t size)
+        : m_size(size)
```
Remove.  See comment below.

MenloDorian commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,238 @@
+    FixedBitVector(size_t size)
+        : m_size(size)
+    {
+        m_bitVector.resize(size);
```
This is not the best way to initialize `m_bitVector`.  The best way is to put it in the initializer list above, and just pass the size to its constructor:
```
    : m_bitVector(size)
```

MenloDorian commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,186 @@
+
+ALWAYS_INLINE bool FixedBitVector::concurrentTestAndSet(size_t bitIndex, Dependency dependency)
+{
+    if (bitIndex >= size())
```
Let's use `UNLIKELY` here.  I think well-behaving code would not be calling with an out of bounds index.

MenloDorian commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,186 @@
+    WordType mask = one << (bitIndex % wordSize);
+    size_t wordIndex = bitIndex / wordSize;
+    WordType* bits = m_bitVector.bits();
+    bool result = bits[wordIndex] & mask;
```
Let's rename `result` to `previousValue`.

MenloDorian commented on Source/WTF/wtf/FixedBitVector.h:
```diff
@@ -0,0 +1,186 @@
+    WordType mask = one << (bitIndex % wordSize);
+    size_t wordIndex = bitIndex / wordSize;
+    WordType* bits = m_bitVector.bits();
+    bool result = bits[wordIndex] & mask;
```
Ditto.

MenloDorian reviewed (approved):
r=me

hyjorc1 committed:
Fix race condition for wasm referenced functions
https://bugs.webkit.org/show_bug.cgi?id=250847
rdar://104332636

Reviewed by Mark Lam.

Fix race condition for updating referenced functions while compiling
wasm functions concurrently.

* JSTests/wasm/stress/referenced-function.js: Added.
(async let):
(j.catch):
* JSTests/wasm/stress/resources/funcref-race.wasm: Added.
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
(JSC::Wasm::ModuleInformation::referencedFunctions const):
(JSC::Wasm::ModuleInformation::hasReferencedFunction const):
(JSC::Wasm::ModuleInformation::addReferencedFunction const):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FixedBitVector.h: Added.
(WTF::FixedBitVector::concurrentAllocateOnce):
(WTF::FixedBitVector::concurrentTest):
(WTF::FixedBitVector::concurrentTestAndSet):
(WTF::FixedBitVector::testAndSet):
(WTF::FixedBitVector::test):
(WTF::FixedBitVector::findBit const):
(WTF::FixedBitVector::operator== const):
(WTF::FixedBitVector::hash const):
(WTF::FixedBitVector::get const):
(WTF::FixedBitVector::dump const):
(WTF::FixedBitVectorHash::hash):
(WTF::FixedBitVectorHash::equal):
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/FixedBitVector.cpp: Added.
(TestWebKitAPI::testFixedBitVectorSize):
(TestWebKitAPI::testFixedBitVectorTest):
(TestWebKitAPI::testFixedBitVectorTestAndSet):
(TestWebKitAPI::testFixedBitVectorConcurrentTest):
(TestWebKitAPI::testFixedBitVectorConcurrentTestAndSet):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/259323@main
